### PR TITLE
[feature/zhs-scraping] Scraping of the ZHS web page

### PR DESCRIPTION
**Pull request by philippzagar**
# Scraping of the ZHS web page

## :recycle: Current situation & Problem
Currently, there is no convenient way to access data of the [ZHS website](https://www.zhs-muenchen.de/zhs/startseite/)

## :bulb: Proposed solution
Build a web scraper with [Cheerio](https://cheerio.js.org/) that scrapes the website data, formats it and puts it into an ordered database structure model

## :gear: Release Notes 
- Add ZHS web scraping

## :heavy_plus_sign: Additional Information
As the ZHS website is quite unreliably, the PR includes multiple fail-safes like automatic request retry and limitation of connections to the website

### Related PRs
None

### Testing
- Insert a scraper object into the Strapi database via the Content Manager (`next_execution_at` and `currentlyRunning` are set via Strapi):
image
- Take a look at the logs to see if new entries are fetched and created/updated
- Check the Content Manager of Strapi to verify that the data is fetched and put into the database correctly

### Reviewer Nudging
`src/scrapers/zhsofferings.js` should be a good starting point. Take a look at how the whole scraping process is done as well as how the scraper is triggered

### CODE REVIEW

**Review by Bentipa — COMMENTED** [review]
Sind nur paar Sachen wie ich sie machen würde.
Ansonsten hab ich nur grob rüber geschaut

File: src/scrapers/utils/axios.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+const axios = require('axios');
+const axiosRetry = require('axios-retry');
+const http = require('http');
+
+const MAX_REQUESTS_COUNT = 3    // Max number of retries, set to 3 as ZHS website is not stable
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply src/scrapers/utils/axios.js]: I included a linter with the "most famous" linting standard, the airbnb one, they include semicolons :D

File: src/scrapers/utils/axios.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+const INTERVAL_MS = 100
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply src/scrapers/utils/axios.js]: added some docs

File: src/scrapers/utils/axios.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+let PENDING_REQUESTS = 0
+
+// create new axios instance
+const instance = axios.create({
+    baseURL: '/',
+    timeout: 8000,
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply src/scrapers/utils/axios.js]: in this case i think configs etc. are overkill. sure u could do that via env variables or smt. but i'm not sure if strapi offers a configuration architecture out of the box

File: src/scrapers/utils/axios.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * Axios Retry Interceptor
+ */
+axiosRetry(instance, {
+    retries: 10,
+    retryDelay: (retryCount, error) => {
+        const retryDelay = (error.message === 'ETIMEDOUT') ? 15000 : (2000 * retryCount)
+        console.log(`Retry attempt: ${retryCount}`)
+        console.log(`Retry delay: ${retryDelay} ms`)
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply src/scrapers/utils/axios.js]: As we don't have a proper observability system right now, i would use these logs as an "error indicator". For sure not a good solution but sure as hell the easiest one

File: src/scrapers/zhsofferings.js [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+
+        // Iterate over every sport type
+        // TODO: Skip the "Restplätze" sport type
+        await Promise.all(sportLinks.map(async (sportLink, i) => {
+            let res = await instance.get(sportLink);
+
+            let $ = cheerio.load(res.data);
+
+            let sportHeadingGerman = $('.bs_head').text()
+            console.log(sportHeadingGerman)
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply src/scrapers/zhsofferings.js]: thx for the catch

File: src/scrapers/zhsofferings.js [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+            let sportHeadingEnglish = $('.bs_kursbeschreibung .bslang_en span strong').map((i, element) => {
+                if ($(element).parent().css('font-size') == '14px') {
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply src/scrapers/zhsofferings.js]: yep that's how shitty the ZHS website is, one need's to differentiate with pixel sizes as well as strong tags :D

File: src/scrapers/utils/axios.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bentipa) [reply src/scrapers/utils/axios.js]: then scrap that :D

File: src/scrapers/utils/axios.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Bentipa) [reply src/scrapers/utils/axios.js]: easy-peasy lemon squeezy, passt

**Review by Bentipa — APPROVED** [review]
wird scho basn

**Pull request merged**